feat(access): Add retirement flag for org roles

RyanSkonnord opened the pull request:
Broken out from https://github.com/getsentry/sentry/pull/32498. Follows https://github.com/getsentry/sentry/pull/33443.

    Add an `is_retired` attribute to organization-level roles. Roles marked
    as retired will be hidden in the UI, and the role can no longer be
    assigned to members. Members who already have the role will keep it,
    receiving the associated permissions and minimum team role.
    
    Role retirement is gated by the "organizations:team-roles" feature flag.
    Organizations without the flag will see retired roles as normal.

nhsiehgit commented on src/sentry/api/endpoints/organization_member/details.py:
```diff
@@ -251,6 +263,11 @@ def put(
+            if organization_roles.get(result["role"]).is_retired and features.has(
+                "organizations:team-roles", organization
+            ):
+                return Response({"detail": "This role is no longer in use."}, status=400)
```
nit: maybe f"role {result['role']} has been deprecated and is no longer in use" ?

I think including the role name in the response might be a good idea 🤷

RyanSkonnord replied:
👍 Replaced with a more descriptive message.

Part of the reason I avoided using the role name in the first place is that the other messages avoided dropping values in, perhaps intentionally. (For example, "You do not have permission to assign the given role.") In other places this is done for ease of localization. I don't think that applies to API messages, but it may be so that API client programs can do direct equality checks against the `"detail"` value. Does anyone know about this?

nhsiehgit reviewed (commented):
+1 lgtm

RyanSkonnord committed:
feat(access): Add retirement flag for org roles

Add an `is_retired` attribute to organization-level roles. Roles marked
as retired will be hidden in the UI, and the role can no longer be
assigned to members. Members who already have the role will keep it,
receiving the associated permissions and minimum team role.

Role retirement is gated by the "organizations:team-roles" feature flag.
Organizations without the flag will see retired roles as normal.

wedamija reviewed (commented):
Could we add a test that creates a retired role and verifies that we return the expected response?

RyanSkonnord committed:
More specific message on assigning to retired role

RyanSkonnord committed:
Extract variable for role attribute

RyanSkonnord committed:
Add test for listing retired roles

RyanSkonnord committed:
Add test_cannot_update_with_retired_role

RyanSkonnord commented:
> Could we add a test that creates a retired role and verifies that we return the expected response?

Done. (https://github.com/getsentry/sentry/pull/33603/commits/e5383d298616242b9c31bba39456358d780b01c7)

wedamija commented on src/sentry/api/endpoints/organization_member/details.py:
```diff
@@ -248,10 +261,16 @@ def put(
                 return Response({"detail": "You cannot make changes to your own role."}, status=400)
 
-            self._change_org_member_role(member, result["role"])
+            if organization_roles.get(assigned_role).is_retired and features.has(
+                "organizations:team-roles", organization
+            ):
```
If the role is the role the user already has, should we just no-op here instead of erroring?

RyanSkonnord committed:
Merge branch 'master' into org-role-retirement

RyanSkonnord committed:
Suppress error response if role is already retired
